src/lib/provable/gadgets/scalar.ts

```diff
@@ -0,0 +1,114 @@
+import type { Field } from '../field.js';
+import type { Bool } from '../bool.js';
+import { Fq } from '../../../bindings/crypto/finite-field.js';
+import { PallasAffine } from '../../../bindings/crypto/elliptic-curve.js';
+import { fieldToField3 } from './comparison.js';
+import { Field3, ForeignField } from './foreign-field.js';
+import { exists, existsOne } from '../core/exists.js';
+import { bit, isConstant, packBits } from './common.js';
+import { TupleN } from '../../util/types.js';
+import { l } from './range-check.js';
+import { createField } from '../core/field-constructor.js';
+import { Snarky } from '../../../snarky.js';
+import { Provable } from '../provable.js';
+import { Group } from '../group.js';
+
+export { scale, scaleShiftedSplit5 };
+
+/**
+ * Gadget to scale a point by a scalar, where the scalar is represented as a _native_ Field.
+ */
+function scale(P: { x: Field; y: Field }, s: Field): Group {
+  // constant case
+  let { x, y } = P;
+  if (x.isConstant() && y.isConstant() && s.isConstant()) {
+    let sP = PallasAffine.scale(
+      PallasAffine.fromNonzero({ x: x.toBigInt(), y: y.toBigInt() }),
+      s.toBigInt()
+    );
+    return new Group({ x: createField(sP.x), y: createField(sP.y) });
+  }
+
+  // compute t = s - 2^254 mod q using foreign field subtraction
+  let sBig = fieldToField3(s);
+  let twoTo254 = Field3.from(1n << 254n);
+  let [t0, t1, t2] = ForeignField.sub(sBig, twoTo254, Fq.modulus);
+
+  // split t into 250 high bits and 5 low bits
+  // => split t0 into [5, 83]
+  let tLo = exists(5, () => {
+    let t = t0.toBigInt();
+    return [bit(t, 0), bit(t, 1), bit(t, 2), bit(t, 3), bit(t, 4)];
+  });
+  let tLoBools = TupleN.map(tLo, (x) => x.assertBool());
+  let tHi0 = existsOne(() => t0.toBigInt() >> 5n);
+
+  // prove split
+  // since we know that t0 < 2^88, this proves that t0High < 2^83
```

Comment:
tHi0 you mean?

Reply:
I'm afraid you're reviewing outdated code :O 
yeah I meant tHi0

Reply:
yeah, yesterday I was commit-reviewing to gain more context. Today I will focus on the updated `native-curve.ts` file